Add `image_diff` comparison method for test output verification using images

kostrykin opened the pull request:
As part of the Imaging Hackathon in Freiburg.

This PR adds the `image_diff` comparison method for test output verification using images.

Previous comparison methods comprise `diff`, `re_match`, `lines_diff`, `re_match_multiline`, `contains`, and `sim_size`, all of which are either unsuitable for image data, or dangerous to be used (because the size of a file does not really represent its content). This PR adds the `image_diff` comparison method, which is specifically tailored for single-channel and multi-channel image data (e.g., RGB), and can be used like this:

```xml
<test>
    <param name="in" value="actual.png" />
    <output name="out" value="expected.tif" compare="image_diff" metric="mae" />
</test>
```

The difference of a pair of images is quantified as the pixel-wise distance between the images, for which different *metrics* can be used. For different applications and different types of images (e.g., intensity vs. label images), different types of metrics are common. The following metrics are added for *intensity images:*

| `metric` | Corresponding metric (for intensity images) |
| --- | --- |
| `mae` | Mean absolute error |
| `mse` | Mean squared error |
| `rms` | Root mean squared error |
| `fro` | Frobenius norm |

For label images (label maps), it is usually irrelevant which particular label is used for labeling an image region, but only the *sets of pixels* corresponding to individual image regions need to be compared. This can be accomplished using *one minus* the *intersection over  the union* of the regions:

| `metric` | Corresponding metric (for label images) |
| --- | --- |
| `iou` | One minus the intersection over the union |

To decide whether a pair of images is equal, a tolerance is required, because different compression algorithms (or different versions thereof) can lead to different compression artifacts. A pair of images is considered to be *equal* in terms of the specified metric, if the distance of the images computed with respect to the metric is not above a threshold `eps` (0.01 by default).

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

kostrykin committed:
Add "image_diff" comparison for test files

kostrykin committed:
Add `files_image_diff`

kostrykin committed:
Add tests for `files_image_diff`

kostrykin committed:
Add `pillow` to pinned-requirements.txt

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix pillow imports

kostrykin committed:
Add pillow dependency

kostrykin committed:
Fix pillow usage

kostrykin committed:
Fix bugs

kostrykin committed:
Fix bugs

kostrykin committed:
Fix bugs

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix bug

kostrykin committed:
Set default `eps` to 0.01

kostrykin committed:
Fix bug

kostrykin committed:
Update XSD

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Add functional test for `image_diff`

kostrykin committed:
Fix functional test for `image_diff` comparison

kostrykin committed:
Add error handler

kostrykin committed:
Update functional test

kostrykin committed:
Fix bugs

kostrykin committed:
Fix bugs

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

bernt-matthias reviewed (commented):
This looks really good to me. Will try to review in the next days.

If I could wish die anything, then a few linting rules for the new attributes might be cool. https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/linters/tests.py

My initial idea was more in the direction of content assertions, e.g. widht, height, colors,... but with the new dependency this should be easy. Could/should go to to a separate PR or issue.

kostrykin committed:
Fix spelling

kostrykin committed:
Add linting for `image_diff`

kostrykin committed:
Remove unrequired test file

nsoranzo committed:
Add type annotations

nsoranzo committed:
Add missing package requirements

jmchilton commented on packages/tool_util/setup.cfg:
```diff
@@ -36,8 +36,10 @@ install_requires =
     galaxy-util>=22.1
     conda-package-streaming
     lxml!=4.2.2
+    numpy
```
numpy is a scary dependency and not required to parse Galaxy XML, etc.. I think we should move these to a new ``extra_require`` - maybe ``extended-assertions``.

nsoranzo committed:
Move numpy and pillow requirements to new extended-assertions extra

together pysam, which is also conditionally imported in
``lib/galaxy/tool_util/verify/__init__.py`` .
